Sources/MessagingPush/RichPush/NSEOperationCoordinator.swift

```diff
@@ -0,0 +1,183 @@
+        logger.debug("Starting coordinated NSE operations for deliveryId: \(String(describing: push.cioDelivery?.id))")
+
+        // Start both operations concurrently
+        startMetricsTracking()
+        startRichPushProcessing()
+    }
+
+    func handleTimeWillExpire() {
+        logger.info("NSE time will expire. Stopping all operations and using current content.")
+
+        // Cancel any pending operations
+        richPushHandler.stopAll()
+
+        // Force completion with whatever content we have
+        let finalContent = state.forceComplete()
+        callContentHandler(with: finalContent, reason: "timeWillExpire")
+    }
+
+    /// Current notification content. Useful for testing thread safety.
+    var currentContent: UNNotificationContent {
+        state.currentContent
+    }
+
+    // MARK: - Private Methods
+
+    private func startMetricsTracking() {
+        // Note: push.cioDelivery is guaranteed to exist because coordinator is only created
+        // when the extension has already verified cioDelivery exists
+        let deliveryInfo = push.cioDelivery!
+
+        logger.debug("Starting metrics tracking for deliveryId: \(deliveryInfo.id)")
+
+        pushDeliveryTracker.trackMetric(token: deliveryInfo.token, event: .delivered, deliveryId: deliveryInfo.id, timestamp: nil) { [weak self] result in
+            self?.handleMetricsCompletion(result: result)
+        }
+    }
+
+    private func startRichPushProcessing() {
+        logger.debug("Starting rich push processing")
+
+        richPushHandler.startRequest(push: push) { [weak self] modifiedPush in
+            self?.handleRichPushCompletion(modifiedPush: modifiedPush)
+        }
+    }
+
+    private func handleMetricsCompletion(result: Result<Void, HttpRequestError>) {
+        switch result {
+        case .success:
+            logger.debug("Metrics tracking completed successfully for deliveryId: \(String(describing: push.cioDelivery?.id))")
+        case .failure(let error):
+            logger.error("Metrics tracking failed for deliveryId: \(String(describing: push.cioDelivery?.id)): \(error)")
+        }
+
+        if let finalContent = state.markMetricsComplete() {
+            callContentHandler(with: finalContent, reason: "bothOperationsComplete")
+        }
+    }
+
+    private func handleRichPushCompletion(modifiedPush: PushNotification) {
+        logger.debug("Rich push processing completed for deliveryId: \(String(describing: push.cioDelivery?.id))")
+
+        let content = (modifiedPush as? UNNotificationWrapper)?.notificationContent ?? state.currentContent
+
+        if let finalContent = state.markRichPushComplete(with: content) {
+            callContentHandler(with: finalContent, reason: "bothOperationsComplete")
+        }
+    }
+
+    private func callContentHandler(with content: UNNotificationContent, reason: String) {
+        logger.info("NSE operations complete (\(reason)). Calling content handler.")
+
+        contentHandler(content)
```

Comment:
should this be done on main thread? did you check with wrappers? they often require NSE operation to be done on main